### PR TITLE
[FLINK-11731] [State Backends] Make DefaultOperatorStateBackend follow the builder pattern

### PR DESCRIPTION
**Pull request by carp84**
## What is the purpose of the change

This PR makes `DefaultOperatorStateBackend` construction follow the builder pattern.

## Brief change log

This PR introduces a builder for `DefaultOperatorStateBackend` and moves all necessary initialization work including restore operation to its building process. It also removes invocation of `DefaultOperatorStateBackend.restore`.

## Verifying this change

This change is already covered by existing tests, such as `OperatorStateBackendTest` and `BackendRestorerProcedureTest`, and verified locally.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes**)
     * It changes the `StateBackend.createOperatorStateBackend` interface, adding two parameters for state handles and `CloseableRegistry` which will be used in restore operation during backend building
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by carp84** [comment]
Checked the travis build and confirmed the failure is irrelative with change here. And a [local travis check](https://travis-ci.org/carp84/flink/builds/501881135) with the same commit passed

**Comment by StefanRRichter** [comment]
Changes to this point look good to me now 👍

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/BackendRestorerProcedure.java [hunk 1]
```diff
@@ -50,7 +49,7 @@
  * @param <S> type of the supplied snapshots from which the backend restores.
  */
 public class BackendRestorerProcedure<
-	T extends Closeable & Disposable & Snapshotable<?, Collection<S>>,
+	T extends Closeable & Disposable & SnapshotStrategy<?>,
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/BackendRestorerProcedure.java]: We can remove the `& SnapshotStrategy<?>` part here, becaues we don't invoke anymore methods from that interface.

**Comment by carp84** [comment]
Thanks for the review and help @StefanRRichter !

**Pull request closed without merging**